### DISCUSSION TITLE
[问题]: Umijs 如何动态话配置router 的排列顺序呢 ?

### DISCUSSION THREAD

**Discussion by Dssssds** [umijs/umi#12177] [General]
# 问题
* umi4, 使用了 getInitialState 和 access.ts 进行了 router 的权限控制.
* 但是如何进行, router 侧边栏排列顺序的自定义呢. 

# 最终效果如图:
![image](https://github.com/umijs/umi/assets/17268259/79f9ff29-4256-4b09-9b3d-47782cdab35f)

**Comment by fz6m** [comment]
在 `.umirc.ts` 配置文件里定义 [`routes`](https://umijs.org/docs/guides/routes) ，使用配置式路由可以自定义菜单顺序。

如果想在运行时更改菜单顺序，可以使用动态路由 [`patchClientRoutes`](https://umijs.org/docs/api/runtime-config#patchclientroutes) 运行时配置：在应用启动时，获取这个角色的权限和拥有的路由表后，动态添加路由，此时就可以排序了。更多内容可参考 #7911 。这种方式是只能有一次在应用启动时构造路由表并排序的机会。

如果想在运行中动态更改顺序，目前 layout 自动生成菜单这个界面是使用了 [ProLayout](https://procomponents.ant.design/components/layout) ，如果有更高的自定义需求，请关闭 layout 插件，自己实现自己的 layout 即可，路由表信息参考 `src/.umi/plugin-layout/Layout.tsx` 来复刻。

**Reply by Dssssds** [reply]
好的好的,感谢.
